### PR TITLE
HIVE-29610: Add support for SQL wildcards (% and _) in SHOW COLUMNS aligning with other SHOW commands

### PR DESCRIPTION
**Pull request by Aggarwal-Raghav**
### What changes were proposed in this pull request?
Check [HIVE-29610](https://issues.apache.org/jira/browse/HIVE-29610)

### Why are the changes needed?
`ShowColumnsOperation` doesn't make use of `UDFLike#likePatternToRegExp()` and only resolve `*` regex pattern and not the `%` which was introduced in [HIVE-23359](https://issues.apache.org/jira/browse/HIVE-23359).

### Does this PR introduce _any_ user-facing change?
Yes, show columns will start supporint query like `SHOW COLUMNS FROM tbl LIKE 'name_%';` and will be backward compatible to support `*` . 

### How was this patch tested?
Running the following UT and cluster testing
```
mvn clean test -Dtest=TestMiniLlapLocalCliDriver -Dqfile=show_columns_inconsistency.q,show_columns.q -Drat.skip -Pitests -pl itests/qtest
```

**Before**
Screenshot 2026-05-14 at 12 41 08 AM

**After**
Screenshot 2026-05-14 at 12 36 58 AM

### CODE REVIEW

**Comment by Aggarwal-Raghav** [comment]
After [HIVE-23359](https://issues.apache.org/jira/browse/HIVE-23359) the behaviour in Hive3 vs Hive4.0.1 for regex pattern was changed i.e. **breaking change** until HIVE-28292 was merged. The following overloaded method where `literalize=false` and `greedyMatch=true` there the support for `*` was restored. Once such example where breaking change still exists is `ShowDatabasesOperation`.
```
public static String likePatternToRegExp(String likePattern, boolean literalize, boolean greedyMatch)
```

In `show_coloums.q` `SHOW SORTED COLUMNS in foo_n7 from test_db "col+";` the `+` doesn't work and i guess it never worked.

**Comment by Aggarwal-Raghav** [comment]
CC @wecharyu @deniskuzZ , requesting for review and feedback on the approach.

**Comment by Aggarwal-Raghav** [comment]
UT failure is un-related. will re-run post review comments

File: ql/src/java/org/apache/hadoop/hive/ql/udf/UDFLike.java [hunk 1]
```diff
@@ -85,6 +85,8 @@ public static String likePatternToRegExp(String likePattern, boolean literalize,
         sb.append(".");
       } else if (n == '%') {
         sb.append(greedyMatch ? ".*" : ".*?");
+      } else if (n == '*' && !literalize) {
+        sb.append(greedyMatch ? ".*" : ".*?");
```
Comment (soumyakanti3578) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/UDFLike.java]: Maybe combine these together?
Also add a small comment explaining the condition

File: ql/src/test/queries/clientpositive/show_columns_inconsistency.q [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  id_primary INT,
+  id_secondary INT,
+  name_first STRING,
+  name_last STRING
+);
+
+
+SHOW TABLES LIKE 'wild%';
+SHOW TABLES LIKE 'wild*';
+SHOW TABLES LIKE 'none|wildcard_table';
```
Comment (soumyakanti3578) [inline ql/src/test/queries/clientpositive/show_columns_inconsistency.q]: These seem out of scope as this PR is fixing SHOW COLUMNS ... LIKE.

File: ql/src/test/queries/clientpositive/show_columns_inconsistency.q
Comment (soumyakanti3578) [inline ql/src/test/queries/clientpositive/show_columns_inconsistency.q:1]: nit: A more relevant name for this? `show_columns_like.q` ?

File: ql/src/test/queries/clientpositive/show_columns_inconsistency.q [hunk 3]
```diff
@@ -0,0 +1,21 @@
+CREATE DATABASE IF NOT EXISTS col_test_db;
+USE col_test_db;
+
+CREATE TABLE wildcard_table (
+  id_primary INT,
+  id_secondary INT,
+  name_first STRING,
+  name_last STRING
+);
```
Comment (soumyakanti3578) [inline ql/src/test/queries/clientpositive/show_columns_inconsistency.q]: Can we add columns here to tests case insensitivity? Since we are using `Pattern.CASE_INSENSITIVE`

File: ql/src/test/queries/clientpositive/show_columns_inconsistency.q [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+
+SHOW TABLES LIKE 'wild%';
+SHOW TABLES LIKE 'wild*';
+SHOW TABLES LIKE 'none|wildcard_table';
+
+SHOW COLUMNS FROM wildcard_table LIKE 'id%';
+SHOW COLUMNS FROM wildcard_table LIKE 'name_%';
```
Comment (soumyakanti3578) [inline ql/src/test/queries/clientpositive/show_columns_inconsistency.q]: Here we are really testing `%`, and not `_`. Can we add more tests for the single char wildcard?
Also add more tests for empty results, like `abc%`, `id__`, for example.

**Review by soumyakanti3578 — COMMENTED** [review]
Overall this looks quite good, but needs cleaning up in a few places and also needs more tests.

**Comment by Aggarwal-Raghav** [comment]
Thanks for the review @soumyakanti3578 , I've addressed the review comments.

File: ql/src/test/queries/clientpositive/show_columns_like.q [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+);
+
+SHOW COLUMNS FROM wildcard_table LIKE 'id%';
+SHOW COLUMNS FROM wildcard_table LIKE 'name_%';
+-- Case Insensitivity test
+SHOW COLUMNS FROM wildcard_table LIKE 'mixedcase%';
+SHOW COLUMNS FROM wildcard_table LIKE 'another_mixed_col';
+SHOW COLUMNS FROM wildcard_table LIKE 'id*';
+SHOW COLUMNS FROM wildcard_table LIKE 'id_primary|name_first';
+
```
Comment (soumyakanti3578) [inline ql/src/test/queries/clientpositive/show_columns_like.q:21]: I think we still need to test escape sequences. For example, add another column to the table - `idxprimary` or something similar, and test with 
```
SHOW COLUMNS FROM wildcard_table LIKE 'id\_primary';
```
which should return just `id_primary` and not `idxprimary`.

File: ql/src/test/queries/clientpositive/show_columns_inconsistency.q (hunk 2 above)
Comment (soumyakanti3578) [reply ql/src/test/queries/clientpositive/show_columns_inconsistency.q]: These tests were out of place in the show columns test file. You can add these in an appropriate file for show tables if these are missing.

**Comment by Aggarwal-Raghav** [comment]
Have addressed the latest review comments as well @soumyakanti3578

**Review by soumyakanti3578 — APPROVED** [review]
LGTM 👍🏼
